src/AspNetCoreModuleV2/CommonLib/AppOfflineApplication.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#include "AppOfflineApplication.h"
+
+#include <experimental/filesystem>
+#include "HandleWrapper.h"
+
+HRESULT AppOfflineApplication::CreateHandler(IHttpContext* pHttpContext, IREQUEST_HANDLER** pRequestHandler)
+{
+    try
+    {
+        *pRequestHandler = new AppOfflineHandler(pHttpContext, m_strAppOfflineContent);
+    }
+    CATCH_RETURN();
+
+    return S_OK;
+}
+
+HRESULT AppOfflineApplication::OnAppOfflineFound()
+{
+    LARGE_INTEGER   li = {};
+
+    HandleWrapper<InvalidHandleTraits> handle = CreateFile(m_appOfflineLocation.c_str(),
+                        GENERIC_READ,
+                        FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
```

Comment:
It appears the .NET Framework 4 app_offline code only sets `FileShare.Read`:

https://referencesource.microsoft.com/#System.Web/HttpRuntime.cs,1254

I believe this will lock the file while it's being read, which would prevent a publish from modifying or deleting the file.  However, this might be fine since publish should retry failed operations, and the file read should be very quick (a few ms).  I would match the behavior of .NET Framework 4, unless we have a good reason to use different behavior with ANCM.

Reply:
It would be better for us not to read the content than to lock the file.

Reply:
Adding a test where `app_offline` file is locked would be nice.

Reply:
>  However, this might be fine since publish should retry failed operations

Considering the number of problems we have even with having publish retry failed operations I would prefer to err on the side of caution.

Reply:
Do we know if the .NET Framework technique of locking the file during read has caused real-world problems for customers?  If not, I hesitate to make the ANCM code more complicated to avoid locking the file.  If we do avoid locking, one invariant I think we must hold is that we either use the entire content of some version of the app_offline file, or no content.  We never use partial content.

For example, say app_offline v1 is 100 bytes, then it's overwritten by a larger file of 200 bytes between the calls to `CreateFile()` and `ReadFile()`.  I believe we might end up serving the first 100 bytes of app_offline v2, which IMO is unacceptable.  I will test this specific scenario in a standalone repro.